nexus/db-model/src/vpc_firewall_rule.rs

```diff
@@ -253,15 +255,40 @@ impl VpcFirewallRule {
     pub fn vec_from_params(
         vpc_id: Uuid,
         params: external::VpcFirewallRuleUpdateParams,
-    ) -> Vec<VpcFirewallRule> {
-        params
+    ) -> Result<Vec<VpcFirewallRule>, external::Error> {
+        ensure_no_duplicates(&params)?;
+        Ok(params
             .rules
             .iter()
             .map(|rule| VpcFirewallRule::new(Uuid::new_v4(), vpc_id, rule))
-            .collect()
+            .collect())
     }
 }
 
+fn ensure_no_duplicates(
+    params: &external::VpcFirewallRuleUpdateParams,
+) -> Result<(), external::Error> {
+    // we could do this by comparing set(names).len() to names.len(), but this
+    // way we can say what the duplicate names are, and that's nice!
+    let mut names = HashSet::new();
+    let mut dupes = HashSet::new();
+    for r in params.rules.iter() {
+        if !names.insert(r.name.clone()) {
+            // insert returns false if already present
+            dupes.insert(r.name.clone());
+        }
+    }
+
+    if !dupes.is_empty() {
+        return Err(external::Error::invalid_value(
+            "rules",
+            format!("Rule names must be unique. Duplicates: {}", json!(dupes)),
```

Comment:
it feels a bit odd to me that we're stuffing JSON into what looks like a human-readable string. does the console actually parse the JSON from this message?

Reply:
...oh, I see, it looks like the `json!` macro is just used as a convenient way to format a comma-delimited list. that's clever --- but maybe it's worth commenting on?

Reply:
I'll see what the more literal map and join equivalent looks like and I'll go with that if it's not too ugly, otherwise will comment.